File: 0516-longest-palindromic-subsequence/0516-longest-palindromic-subsequence.cpp
class Solution {
public:
    int longestPalindromeSubseq(string s) {
        string reverse = "";

        for(int i=s.size()-1; i>=0; i--)
            reverse+=s[i];

        vector<vector<int>> table (s.size()+1, vector<int>(s.size()+1,0));

        for(int i=1; i<s.size()+1; i++){
            for(int j=1; j<s.size()+1; j++){
                if(s[i-1]==reverse[j-1])
                    table[i][j] = 1 + table[i-1][j-1];
                else
                    table[i][j] = max(table[i-1][j], table[i][j-1]);
            }
        }
        return table[s.size()][s.size()];

    }
};

File: 0516-longest-palindromic-subsequence/0516-longest-palindromic-subsequence.java
class Solution {
    public int longestPalindromeSubseq(String s) {
        
        int len = s.length();
        String revString = (new StringBuilder(s)).reverse().toString();

        int[][] dp = new int[len+1][len+1];

        int[] prev = new int[len+1];
        

        for(int i=1; i<len+1; i++){
            int[] curr = new int[len+1];
            for(int j=1; j<len+1; j++){

                if(s.charAt(i-1)==revString.charAt(j-1))
                    curr[j] = 1+prev[j-1];
                else
                    curr[j] = Math.max(curr[j-1], prev[j]);

            }
            prev=curr;
        }

        return prev[len];
    }
}